cli/command/stack/deploy_test.go

```diff
@@ -22,3 +25,84 @@ func TestPruneServices(t *testing.T) {
 	pruneServices(ctx, dockerCli, namespace, services)
 	assert.Equal(t, buildObjectIDs([]string{objectName("foo", "remove")}), client.removedServices)
 }
+
+// TestServiceUpdateResolveImageChanged tests that the service's
+// image and placement are preserved if the image did not change
+func TestServiceUpdateResolveImageChanged(t *testing.T) {
+	ctx := context.Background()
+	namespace := convert.NewNamespace("mystack")
+
+	type result struct {
+		queryRegistry bool
+		image         string
+		platform      swarm.Platform
+	}
+	var expected result
+
+	client := test.NewFakeCli(&fakeClient{
+		serviceListFunc: func(options types.ServiceListOptions) ([]swarm.Service, error) {
+			return []swarm.Service{
+				{
+					Spec: swarm.ServiceSpec{
+						Annotations: swarm.Annotations{
+							Name:   namespace.Name() + "_myservice",
+							Labels: map[string]string{"com.docker.stack.image": "foobar:1.2.3"},
+						},
+						TaskTemplate: swarm.TaskSpec{
+							ContainerSpec: swarm.ContainerSpec{
+								Image: "foobar:1.2.3@sha256:deadbeef",
+							},
+							Placement: &swarm.Placement{
+								Platforms: []swarm.Platform{
+									{
+										Architecture: "foo-architecture",
+										OS:           "foo-os",
+									},
+								},
+							},
+						},
+					},
+				},
+			}, nil
+		},
+		serviceUpdateFunc: func(serviceID string, version swarm.Version, service swarm.ServiceSpec, options types.ServiceUpdateOptions) (types.ServiceUpdateResponse, error) {
+			assert.Equal(t, expected.queryRegistry, options.QueryRegistry)
+			assert.Equal(t, expected.image, service.TaskTemplate.ContainerSpec.Image)
+			if !expected.queryRegistry {
+				require.NotNil(t, service.TaskTemplate.Placement)
+				require.Len(t, service.TaskTemplate.Placement.Platforms, 1)
+				assert.Equal(t, expected.platform, service.TaskTemplate.Placement.Platforms[0])
+			}
+			return types.ServiceUpdateResponse{}, nil
+		},
+	})
+
+	tests := map[string]result{
+		// Image not changed
+		"foobar:1.2.3": {
+			queryRegistry: false,
+			image:         "foobar:1.2.3@sha256:deadbeef",
+			platform:      swarm.Platform{Architecture: "foo-architecture", OS: "foo-os"},
+		},
+		// Image changed
+		"foobar:1.2.4": {
+			queryRegistry: true,
+			image:         "foobar:1.2.4",
+		},
+	}
+
+	for image, res := range tests {
+		t.Logf("Testing image %q", image)
+		expected = res
+		spec := map[string]swarm.ServiceSpec{
+			"myservice": {
+				TaskTemplate: swarm.TaskSpec{
+					ContainerSpec: swarm.ContainerSpec{
+						Image: image,
+					},
+				},
+			},
+		}
+		deployServices(ctx, client, spec, namespace, false, resolveImageChanged)
```

Comment:
This test would continue to pass if `serviceUpdateFunc()` was not called, even if the test was failing. I think we need to adjust the assertions.

`deployServices()` can return an error, so we should `assert.NoError(t, err)` on that error.

Maybe instead of asserting in the fake, the fake could store the arg values, and the assertions could happen here after the call to `deployServices()` ?